src/marked.js

```diff
@@ -139,44 +139,98 @@ marked.defaults = defaults;
  */
 
 marked.use = function(extension) {
-  const opts = merge({}, extension);
-  if (extension.renderer) {
-    const renderer = marked.defaults.renderer || new Renderer();
-    for (const prop in extension.renderer) {
-      const prevRenderer = renderer[prop];
-      renderer[prop] = (...args) => {
-        let ret = extension.renderer[prop].apply(renderer, args);
-        if (ret === false) {
-          ret = prevRenderer.apply(renderer, args);
+  if (!Array.isArray(extension)) { // Wrap in array if not already to unify processing
+    extension = [extension];
+  }
+  const opts = merge({}, ...extension);
+  opts.tokenizer = null;
+  opts.renderer = null;
+  opts.extensions = null;
+  const extensions = { renderers: {}, walkableTokens: {} };
```

Comment:
I added a test that uses multiple extensions in an array and calling `marked.use` multiple times and it looks like different things are getting overwritten depending on how multiple extensions are loaded.

Reply:
> I added a test that uses multiple extensions in an array

I'm working on this, but I see a possible error in your block extension. Just want to make sure I'm testing the extension as it is intended. This:

```
renderer(token) {
  return `<${token.type}>${token.text}</${token.type}>\n`;
}
```
should be this:
```
renderer(token) {
  return `<${token.type}>${this.parseInline(token.tokens)}</${token.type}>\n`;
}
```
Otherwise the inline tokens are generated but never rendered and we get `<block-extension1>:extension1</block-extension1>` instead of `<block-extension1>used extension1 walked</block-extension1>`. Also we probably need to turn off the `headerIds` option because they are being generated in the test which doesn't match the expected output.

Is this what you intended? I want to make sure I'm fixing the right bug.

Reply:
Yes. nice catch

Reply:
As for the extensions that overwrite default tokenizers, I imagine the latest one to be loaded should overwrite the first one. i.e. this:
https://github.com/markedjs/marked/blob/64cdfa2e88131404e077c9e56bbc7425630263ac/test/unit/marked-spec.js#L395-L414
will only apply to `# extension2`  because when extension 2 is loaded, `name` is overwritten and won't match `# extension1`.

So expected output for 
```
# extension1

# extension2

# no extension
```
becomes:
```
<h1>extension1</h1>   // overwritten by extension2
<h1>used extension2 walked</h1>
<h1>no extension</h1>
```

... Or should `# extension1` and `# no extension` even be translated at all? Since the heading tokenizer is completely overwritten to only match `# extension 2`?

Reply:
The heading tokenizer in extension1 shouldn't be overwritten. If extension 2 heading tokenizer returns `false` than it should drop back to extensions 1 heading tokenizer, and if extension 1 heading tokenizer returns false it should drop back to the original heading tokenizer.

Reply:
Oh what! I didn't realize it left previous tokenizers intact. Ok that is good to know.

Reply:
Wait where is the "fallback" logic in the code? Can you point me to it?

Reply:
https://github.com/markedjs/marked/blob/29a301bf8ce5b6bf3524eb49efb0e1e669982303/src/marked.js#L157-L170

Reply:
Ah wait I see I see... I could not tell exactly what that bit of code was doing but now I see it replaces the default tokenizer with a function that first uses the extension and then if it fails, runs again on the previous tokenizer, which itself could be another function depending how many extensions have been loaded that overwrite the same tokenizer.

Oh my gosh this changes everything. I totally missed that this had a fallback baked in. Oh my gosh. This solves so many problems with my own projects. Oh my gosh.

Reply:
Done!